File: python/src/tdcv2/distribution/hamilton.py
"""Turning percentages into whole rows, exactly.

``percent="50,30,20"`` over seven records cannot be met by rounding: the shares come to 3.5, 2.1
and 1.4. Largest-remainder apportionment — the same method used to allocate seats to parties —
hands out the whole parts first and then the leftovers, so the columns always sum to the count
and each share is as close to what was asked for as whole rows allow.

Ties are the interesting part. When several values have the same remainder and there are fewer
leftovers than claimants, the choice is made from the run's own generator rather than by index
order. Breaking ties by index would make the first-declared value quietly win every close call,
and a config that lists its values in a different order would produce different data.
"""

from __future__ import annotations

import math
from typing import TypeVar

from ..prng import rand
from ..prng.prng import Sfc32

T = TypeVar("T")


def counts_per_value(count: int, percents: list[float], prng: Sfc32) -> list[int]:
    """How many rows each value receives."""
    card_percent = 100.0 / count
    counts = [0] * len(percents)
    remainders = [0.0] * len(percents)
    filled = 0

    for i, percent in enumerate(percents):
        raw_cells = percent / card_percent
        whole = int(raw_cells)  # truncation toward zero, as Math.trunc does
        counts[i] = whole
        remainders[i] = raw_cells % 1
        filled += whole

    unallocated = count - filled
    if unallocated <= 0:
        return counts

    # Largest remainder first; index order only to make the grouping below deterministic.
    order = sorted(range(len(remainders)), key=lambda i: (-remainders[i], i))

    at = 0
    while unallocated > 0 and at < len(order):
        remainder = remainders[order[at]]
        end = at
        while end < len(order) and remainders[order[end]] == remainder:
            end += 1

        group_size = end - at
        if group_size <= unallocated:
            for k in range(at, end):
                counts[order[k]] += 1
                unallocated -= 1
            at = end
            continue

        # More claimants than leftovers: draw, rather than let declaration order decide.
        pool = list(order[at:end])
        while unallocated > 0:
            picked = math.floor(prng.next() * len(pool))
            counts[pool[picked]] += 1
            pool.pop(picked)
            unallocated -= 1

    return counts


def distribute(count: int, values: list[T], percents: list[float], prng: Sfc32) -> list[T]:
    """The materialized, shuffled sequence of ``count`` values."""
    counts = counts_per_value(count, percents, prng)
    sequence: list[T] = []
    for i, value in enumerate(values):
        sequence.extend([value] * counts[i])
    return rand.shuffle(prng, sequence)

File: csharp/Tdcv2/Distribution/Hamilton.cs
using Tdcv2.Prng;

namespace Tdcv2.Distribution;

/// <summary>
/// Hamilton's largest-remainder method: split <c>count</c> rows across values in the declared
/// percentages, exactly.
/// </summary>
/// <remarks>
/// <para>
/// This is what makes <c>percent="60,40"</c> produce 60 and 40 rather than "about" 60 and 40.
/// Each value first takes its whole share; the rows left over by rounding go to the largest
/// fractional remainders.
/// </para>
/// <para>Two details decide whether a port matches the reference, and both are easy to get wrong:</para>
/// <list type="number">
///   <item>
///     <b>Tie order.</b> Values with equal remainders are served lowest index first. Only when a
///     tie group is larger than the number of rows left does the generator get consulted, one draw
///     per row, from a pool that shrinks as it goes.
///   </item>
///   <item>
///     <b>Draw accounting.</b> Tie-breaking and the final shuffle both consume from the same
///     generator, in that order. Drawing a different number of times leaves the generator in a
///     different state, and everything generated afterwards diverges — even though the counts
///     themselves would still look correct.
///   </item>
/// </list>
/// <para>Verified against <c>fixtures/cross-language/hamilton-vectors.json</c>.</para>
/// </remarks>
public static class Hamilton
{
    /// <summary>How many rows each value receives.</summary>
    public static int[] CountsPerValue(int count, double[] percents, Sfc32 prng)
    {
        double cardPercent = 100.0 / count;
        var counts = new int[percents.Length];
        var remainders = new double[percents.Length];

        int filled = 0;
        for (int i = 0; i < percents.Length; i++)
        {
            double rawCells = percents[i] / cardPercent;
            int whole = (int)rawCells; // truncation toward zero, as Math.trunc does
            counts[i] = whole;
            remainders[i] = rawCells % 1;
            filled += whole;
        }

        int unallocated = count - filled;
        if (unallocated <= 0)
        {
            return counts;
        }

        // Remainder descending, index ascending — the order the reference walks in.
        int[] order = Enumerable.Range(0, remainders.Length)
            .OrderByDescending(i => remainders[i])
            .ThenBy(i => i)
            .ToArray();

        int at = 0;
        while (unallocated > 0 && at < order.Length)
        {
            double remainder = remainders[order[at]];
            int end = at;
            while (end < order.Length && remainders[order[end]] == remainder)
            {
                end++;
            }

            int groupSize = end - at;
            if (groupSize <= unallocated)
            {
                for (int k = at; k < end; k++)
                {
                    counts[order[k]]++;
                    unallocated--;
                }

                at = end;
                continue;
            }

            // More values tied than rows left: pick one at random per row, from a pool that
            // shrinks with each pick. One draw per row, which is what keeps the generator in step.
            var pool = new List<int>(order[at..end]);
            while (unallocated > 0)
            {
                int pick = (int)Math.Floor(prng.Next() * pool.Count);
                counts[pool[pick]]++;
                pool.RemoveAt(pick);
                unallocated--;
            }
        }

        return counts;
    }

    /// <summary>The materialised, shuffled sequence of <c>count</c> values.</summary>
    public static IReadOnlyList<T> Distribute<T>(
        int count, IReadOnlyList<T> values, double[] percents, Sfc32 prng)
    {
        int[] counts = CountsPerValue(count, percents, prng);
        var sequence = new List<T>(count);
        for (int i = 0; i < values.Count; i++)
        {
            for (int j = 0; j < counts[i]; j++)
            {
                sequence.Add(values[i]);
            }
        }

        return Shuffle(prng, sequence);
    }

    /// <summary>
    /// Fisher-Yates, from the end backwards.
    /// </summary>
    /// <remarks>
    /// The direction is not a detail. Walking the array the other way consumes the same number of
    /// draws but pairs them with different indices, so a port that flips it produces a shuffle
    /// that is equally valid and not the same one.
    /// </remarks>
    public static IReadOnlyList<T> Shuffle<T>(Sfc32 prng, IReadOnlyList<T> values)
    {
        var result = new List<T>(values);
        for (int i = result.Count - 1; i > 0; i--)
        {
            int j = (int)Math.Floor(prng.Next() * (i + 1));
            (result[i], result[j]) = (result[j], result[i]);
        }

        return result;
    }
}
